healthcheck.go

```diff
@@ -0,0 +1,269 @@
+		Value:    task,
+		Priority: time.Now().UTC().Add(time.Duration(after) * time.Millisecond).Unix(),
+	}
+	ts.logger.Debug().Msgf("HC@Offer expiration offered as %d time: %d", after, item.Priority)
+	ts.mu.Lock()
+	heap.Push(&ts.Q, item)
+	ts.mu.Unlock()
+
+	// Check if we actually pushed new element on the top meaning new element is
+	// ready for the dispatch
+	if item.Index == 0 {
+		if atomic.CompareAndSwapInt32(&ts.isSleeping, 1, 0) {
+			// Wake up whole pool of workers
+			ts.taskAdded <- 1
+		}
+	}
+}
+
+func (ts *HealthCheckScheduler) poll(ctx context.Context, pollerId int) (*healthCheckItem, error) {
+	iteratedOnTask := int64(0)
+	for {
+		isNow := time.Now().UTC().Unix()
+		ts.mu.Lock()
+		i := ts.Q.Peek()
+		iteratedOnTask++
+		if i == nil {
+
+			// Set Sleep condition here
+			atomic.StoreInt32(&ts.isSleeping, 1)
+
+			ts.mu.Unlock()
+
+			// Wait for semaphore to unlock
+			select {
+			case <-ts.taskAdded:
+				ts.logger.Debug().Msgf("HC@Poller <%d> Called out on task added", pollerId)
+				break
+			case <-ctx.Done():
+				return nil, fmt.Errorf("context ended for the Poll action")
+			}
+
+			continue
+
+		} else {
+
+			iteratedOnTask++
+			task := i.(*taskItem)
+			// If task is ready to pop
+			if task.Priority <= isNow {
+				i = heap.Pop(&ts.Q)
+				ts.mu.Unlock()
+				ts.logger.Debug().Msgf("HC@Poller <%d> poll dequeue id <%d> at <%d> system <%d> loadIter <%d>", pollerId, task.Id, task.Priority, isNow, iteratedOnTask)
+				return task.Value.(*healthCheckItem), nil
+			} else {
+				// If Task should await for the next moment
+				ts.mu.Unlock()
+				select {
+				// Wait for general condition unlock
+				case <-ts.taskAdded:
+					ts.logger.Debug().Msgf("HC@Poller <%d> Called out on task added", pollerId)
+					continue
+				// Delay next checkup
+				case <-time.After(time.Duration(task.Priority-isNow) * time.Millisecond * 1000):
+					ts.logger.Debug().Msgf("HC@Poller <%d> Called out on time duration block end", pollerId)
+					continue
+				case <-ctx.Done():
+					return nil, fmt.Errorf("context ended for the Poll action")
+				}
+			}
+
+		}
+	}
+}
+
+// Object to operate within queue
+type taskItem struct {
+	Id       int64
+	Value    interface{}
+	Priority int64
+	Index    int
+}
+
+// Min Queue
+type taskQueue []*taskItem
+
+func newTaskQueue(capacity int) taskQueue {
+	return make(taskQueue, 0, capacity)
+}
+
+func (pq taskQueue) Len() int {
+	return len(pq)
+}
+
+func (pq taskQueue) Less(i, j int) bool {
+	return pq[i].Priority < pq[j].Priority
+}
+
+func (pq taskQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].Index = i
+	pq[j].Index = j
+}
+
+func (pq *taskQueue) Push(x interface{}) {
+	n := len(*pq)
+	c := cap(*pq)
+	if n+1 > c {
+		npq := make(taskQueue, n, c*2)
+		copy(npq, *pq)
+		*pq = npq
+	}
+	*pq = (*pq)[0 : n+1]
+	item := x.(*taskItem)
+	item.Index = n
+	(*pq)[n] = item
+}
+
+func (pq *taskQueue) Pop() interface{} {
+	n := len(*pq)
+	c := cap(*pq)
+	if n < (c/2) && c > 25 {
```

Comment:
It's difficult to understand the significance of "25" etc here without a comment explaining this behaviour.

Reply:
Certainly, the number is arbitrary and designates minimal capacity allocation, which should be fixed to a lower value, assuming we are dealing with occasional hosts going down and most cases will be lower range.